include/zephyr/drivers/crc.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CRC_H
+#define ZEPHYR_INCLUDE_DRIVERS_CRC_H
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief CRC driver APIs
+ * @defgroup crc_interface CRC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#define CRC_FLAG_REVERSE_INPUT  BIT(0)
+#define CRC_FLAG_REVERSE_OUTPUT BIT(1)
+
+#define CRC4_INIT_VAL        0x0
+#define CRC4_TI_INIT_VAL     0x0
+#define CRC7_BE_INIT_VAL     0x0
+#define CRC8_INIT_VAL        0x0
+#define CRC8_CCITT_INIT_VAL  0x00
+#define CRC16_INIT_VAL       0x0
+#define CRC16_ANSI_INIT_VAL  0x0
+#define CRC16_CCITT_INIT_VAL 0x0000
+#define CRC16_ITU_T_INIT_VAL 0x0000
+#define CRC24_PGP_INIT_VAL   0x0
+#define CRC32_C_INIT_VAL     0x00000000
+#define CRC32_IEEE_INIT_VAL  0xFFFFFFFFU
+
+/**
+ * @brief CRC algorithm enumeration
+ */
+enum crc_type {
+	CRC4,
+	CRC4_TI,
+	CRC7_BE,
+	CRC8,
+	CRC8_CCITT,
+	CRC16,
+	CRC16_ANSI,
+	CRC16_CCITT,
+	CRC16_ITU_T,
+	CRC24_PGP,
+	CRC32_C,
+	CRC32_IEEE,
+};
+
+/**
+ * @brief CRC state enumeration
+ */
+enum crc_state {
+	CRC_STATE_IDLE,
+	CRC_STATE_IN_PROGRESS,
+	CRC_STATE_COMPLETE
+};
+
+/**
+ * @brief Runtime driver configuration
+ * @param mutex Zephyr mutex object to make driver thread-safe when driver
+ *		is used within multiple threads
+ */
+struct crc_data {
+	struct k_mutex mutex;
+	k_timeout_t timeout;
+};
+
+typedef uint64_t crc_polynomial_t;
+typedef uint64_t crc_result_t;
+
+/**
+ * @brief CRC context structure
+ */
+struct crc_ctx {
+	enum crc_type type;          /**< Type of CRC algorithm */
+	enum crc_state state;        /**< State of CRC calculation */
+	const void *initial_value;   /**< Initial value for CRC calculation */
+	uint32_t flags;              /**< Flags for CRC calculation */
+	crc_polynomial_t polynomial; /**< Polynomial used in CRC calculation */
+	crc_result_t result;         /**< Store the CRC result */
+};
+
+typedef int (*crc_api_lock)(const struct device *dev, k_timeout_t timeout);
+typedef int (*crc_api_unlock)(const struct device *dev);
+typedef int (*crc_api_begin)(const struct device *dev, struct crc_ctx *ctx, enum crc_type type,
+			     uint32_t flags, crc_polynomial_t polynomial,
+			     const void *initial_value);
+typedef int (*crc_api_update)(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			      size_t bufsize);
+typedef int (*crc_api_verify)(const struct device *dev, struct crc_ctx *ctx, crc_result_t expected);
```

Comment:
Why not something like:
```suggestion
typedef int (*crc_api_finish)(const struct device *dev, struct crc_ctx *ctx, crc_result_t *result);
```
Current API unnecessarily restricts use cases. `verify` can be trivially implemented in-header:
```c
int crc_finish_and_verify(const struct device *dev, struct crc_ctx *ctx, crc_result_t expected)
{
   crc_result_t crc_res;
   int res = crc_finish(dev, ctx, &res);
   if (res < 0) {
      return res;
   }

   if (crc_res != expected) {
      return -EINVAL; //or w/e
   }
   return 0;
}
```

Reply:
Would you be able to clarify the intended use of the `result` parameter in the proposed `crc_api_finish` function? Is the goal is to provide direct access to the CRC result without requiring users to manipulate the ctx structure directly, or something else?

Reply:
Yes, exactly. Currently, the `struct crc_ctx` contains the `result` after each `update`, but it makes more sense to only compute `result` once, after all bytes to checksum have been provided via `update`s. This is just a proposal.
In any case, current implementation of `crc_api_verify` in driver should *not* be in the driver because it has nothing hardware-specific.

Reply:
In the current implementation,  the crc_update function handles the full data buffer, processes it, and updates the result in the ctx struct. I not sure I see the functionality of moving the result retrieval to the finish function

Reply:
maybe it would not be necessary to return CRC result in `finish`, but in any case, if we want to follow the crypto API, we need a `begin` and `finish`